src/Elastic.Apm/Model/LabelsDictionary.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System.Collections.Generic;
+using Elastic.Apm.Api;
+
+namespace Elastic.Apm.Model
+{
+	/// <summary>
+	/// The instance of this type holds the <see cref="Dictionary{TKey,TValue}" /> that contains all the labels on a given
+	/// event.
+	/// The public interface of this mimics a Dictionary with <see cref="string" /> as both the key and the value.
+	/// The reason for this is backwards compatibility - this type makes sure that we don't break user that rely on the old
+	/// interface.
+	/// </summary>
+	internal class LabelsDictionary : Dictionary<string, string>
```

Comment:
Originally I thought we can shadow the `int this[string key]` indexing operator with the `new` keyword. Unfortunately this does not help, because the API surface works with `Dictionary<string,string>`.

On `Transaction.cs` we have:

```
public Dictionary<string, string> Labels => Context.Labels;
````

and `Context.Labels` is the instance of this `LabelsDictionary`. 

But if we use the `new` keyword to shadow, then `Transaction.Labels["foo"] = "bar"` will hit the `int this[T key]` operator of `Dictionary` and not the one that we'd implement here. 

So, the only option I could come up with is to have 2 dictionaries: 1 for the items used through the `Dictionary<string,string>` API and 1 for the ones added through the `SetLabel` method - and we merge them before serialization.